File: bnc/src/main/java/org/sqlbuilder/bnc/BNCUpdatingProcessor.java
package org.sqlbuilder.bnc;

import org.sqlbuilder.bnc.objects.BNCExtendedResolvingRecord;
import org.sqlbuilder.bnc.objects.BNCRecord;
import org.sqlbuilder.bnc.objects.BNCResolvingRecord;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.util.Properties;
import java.util.function.BiConsumer;

public class BNCUpdatingProcessor extends BNCResolvingProcessor
{
	public BNCUpdatingProcessor(final Properties conf) throws IOException, ClassNotFoundException
	{
		super(conf);
	}

	@Override
	public void run() throws IOException
	{
		// main
		String bNCMain = conf.getProperty("bnc_main", "bnc.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.updateFile("bncs"))), true, StandardCharsets.UTF_8))
		{
			processBNCFile(ps, new File(bncHome, bNCMain), (record, i) -> updateRow(ps, names.table("bncs"), i, record.updateRow(names.get("bncs.wordid"), names.get("bncs.word"))));
		}

		// subfiles
		String bNCSpWr = conf.getProperty("bnc_spwr", "bnc-spoken-written.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.updateFile("spwrs"))), true, StandardCharsets.UTF_8))
		{
			processBNCSubFile(ps, new File(bncHome, bNCSpWr), (record, i) -> updateRow(ps, names.table("spwrs"), i, record.updateRow(names.get("spwrs.wordid"), names.get("spwrs.word"))));
		}

		String bNCConvTask = conf.getProperty("bnc_convtask", "bnc-conv-task.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.updateFile("convtasks"))), true, StandardCharsets.UTF_8))
		{
			processBNCSubFile(ps, new File(bncHome, bNCConvTask), (record, i) -> updateRow(ps, names.table("convtasks"), i, record.updateRow(names.get("convtasks.wordid"), names.get("convtasks.word"))));
		}

		String bNCImagInf = conf.getProperty("bnc_imaginf", "bnc-imag-inf.txt");
		try (PrintStream ps = new PrintStream(new FileOutputStream(new File(outDir, names.updateFile("imaginfs"))), true, StandardCharsets.UTF_8))
		{
			processBNCSubFile(ps, new File(bncHome, bNCImagInf), (record, i) -> updateRow(ps, names.table("imaginfs"), i, record.updateRow(names.get("imaginfs.wordid"), names.get("imaginfs.word"))));
		}
	}

	private void processBNCFile(final PrintStream ps, final File file, final BiConsumer<BNCRecord, Integer> consumer) throws IOException
	{
		ps.printf("-- %s %s%n", file.getName(), this.serFile);
		process(file, BNCResolvingRecord::parse, consumer);
	}

	private void processBNCSubFile(final PrintStream ps, final File file, final BiConsumer<BNCRecord, Integer> consumer) throws IOException
	{
		ps.printf("-- %s %s%n", file.getName(), this.serFile);
		process(file, BNCExtendedResolvingRecord::parse, consumer);
	}

	private void updateRow(final PrintStream ps, final String table, final Integer index, final String updateRow)
	{
		ps.printf("UPDATE `%s` SET %s; -- %d%n", table, updateRow, index + 1);
	}
}

File: bnc/src/main/java/org/sqlbuilder/bnc/BncUpdatingProcessor.kt
package org.sqlbuilder.bnc

import org.sqlbuilder.bnc.objects.BncExtendedRecord
import org.sqlbuilder.bnc.objects.BncRecord
import org.sqlbuilder.common.Utils.escape
import org.sqlbuilder.common.Utils.quote
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.PrintStream
import java.nio.charset.StandardCharsets
import java.util.*

class BncUpdatingProcessor(conf: Properties) : BncResolvingProcessor(conf) {
    init {
        // output
        outDir = File(conf.getProperty("bnc_outdir_updated", "sql/data_updated"))
        if (!outDir.exists()) {
            outDir.mkdirs()
        }
    }

    @Throws(IOException::class)
    private fun processBNCFile(ps: PrintStream, file: File, consumer: (BncRecord, Int) -> Unit) {
        process(file, { BncRecord.Companion.parse(it) }, consumer)
    }

    @Throws(IOException::class)
    private fun processBNCSubFile(ps: PrintStream, file: File, consumer: (BncRecord, Int) -> Unit) {
        process(file, { BncExtendedRecord.Companion.parse(it) }, consumer)
    }

    @Throws(IOException::class)
    override fun run() {
        // main
        val bNCMain = conf.getProperty("bnc_main", "bnc.txt")
        PrintStream(FileOutputStream(File(outDir, names.updateFile("bncs"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCFile(it, File(bncHome, bNCMain), { record: BncRecord, i: Int -> updateRow(it, names.table("bncs"), i, record, names.column("bncs.wordid"), names.column("bncs.word")) })
        }
        // subfiles
        val bNCSpWr = conf.getProperty("bnc_spwr", "bnc-spoken-written.txt")
        PrintStream(FileOutputStream(File(outDir, names.updateFile("spwrs"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCSubFile(it, File(bncHome, bNCSpWr), { record: BncRecord, i: Int -> updateRow(it, names.table("spwrs"), i, record, names.column("spwrs.wordid"), names.column("spwrs.word")) })
        }
        val bNCConvTask = conf.getProperty("bnc_convtask", "bnc-conv-task.txt")
        PrintStream(FileOutputStream(File(outDir, names.updateFile("convtasks"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCSubFile(it, File(bncHome, bNCConvTask), { record: BncRecord, i: Int -> updateRow(it, names.table("convtasks"), i, record, names.column("convtasks.wordid"), names.column("convtasks.word")) })
        }
        val bNCImagInf = conf.getProperty("bnc_imaginf", "bnc-imag-inf.txt")
        PrintStream(FileOutputStream(File(outDir, names.updateFile("imaginfs"))), true, StandardCharsets.UTF_8).use {
            it.println("-- $header")
            processBNCSubFile(it, File(bncHome, bNCImagInf), { record: BncRecord, i: Int -> updateRow(it, names.table("imaginfs"), i, record, names.column("imaginfs.wordid"), names.column("imaginfs.word")) })
        }
    }

    private fun updateRow(ps: PrintStream, table: String, index: Int, bncRecord: BncRecord, vararg columns: String) {
        val wordid: Int? = wordResolver.apply(bncRecord.word)
        if (wordid != null) {
            val setClause = "${columns[0]}=$wordid"
            val whereClause = "${columns[1]}=${quote(escape(bncRecord.word))}"
            ps.println("UPDATE $table SET $setClause WHERE $whereClause; -- ${index + 1}")
        }
    }
}
